[Oneshot Refactor] Main refactor (#1110)
vllm-project/llm-compressor: src/llmcompressor/entrypoints/README.md

# LLM Compressor Entrypoints

## Oneshot

Model optimizations compress models while preserving accuracy. One-shot in LLM-Compressor supports faster inference on vLLM by applying post-training quantization (PTQ) or sparsification

> 

Reviewer: Let's focus this readme on the components/steps in OneShot.  We can take most of those details from the docstring you've added.

- Recommended for any GPUs types. 
#### W8A8-INT8 
- Uses channel-wise quantization to compress weights to 8 bits, and uses dynamic per-token quantization to compress activations to 8 bits. 
- Useful for speed ups in high QPS regimes or offline serving on vLLM. 
- Recommended for NVIDIA GPUs with compute capability <8.9 (Ampere, Turing, Volta, Pascal, or older). 
#### W8A8-FP8
- Uses channel-wise quantization to compress weights to 8 bits, and uses dynamic per-token quantization to compress activations to 8 bits. 
- Useful for speed ups in high QPS regimes or offline serving on vLLM. 
- Recommended for NVIDIA GPUs with compute capability >8.9 (Hopper and Ada Lovelace). 

### Sparsification
Sparsification reduces model complexity by pruning selected weight values to zero while retaining essential weights in a subset of parameters. Supported formats include:

#### 2:4-Sparsity
- Uses semi-structured sparsity, where for every four contiguous weights in a tensor, two are set to zero. 
- Useful for efficiently computing sparse matrix multiplications using NVIDIA Sparse Tensor Cores. To preserve as much accuracy from the base model, usually a sparse-finetune step is added. 

> - Recommended for NVIDIA architecture from Amphere onwards, supported on GPUs with Sparse Tensor Cores

Reviewer: I think this is great but I think we should do this as a follow-up so it's not blocking this PR after.

> - Recommended for NVIDIA architecture from Amphere onwards, supported on GPUs with Sparse Tensor Cores

Reviewer: I actually think cutlass requires min hopper?

Author: Yeah to use the kernels we developed, you are right. 
https://neuralmagic.com/blog/24-sparse-llama-fp8-sota-performance-for-nvidia-hopper-gpus/

> # LLM Compressor Entrypoints

Reviewer 2: This file seems incomplete?

## Oneshot

> Model optimizations compresses models while preserving accuracy. Oneshot in llm-compressor supports faster infernce on vLLM by applying post-training quantization (PTQ) or sparsification.

Reviewer 2: ```suggestion
Model optimizations compresses models while preserving accuracy. Oneshot in llm-compressor supports faster inference on vLLM by applying post-training quantization (PTQ) or sparsification.
```

### PTQ
PTQ is carried out to reduce the precision of quantizable weights, (ex. Linear layers) to a lower bit-width. Supported formats are 

#### W4A16

> - Uses GPTQ for weights to compress weights to 4 bytes. 

Reviewer 2: ```suggestion
- Uses GPTQ for weights to compress weights to 4 bit. 
```

Author: Yes you are correct

- Useful speed ups in low QPS regimes with more weight compression. 
- Recommended for general commodity / enterprize GPUs (NVIDIA A10). 
#### W8A8-int8 
- Uses channel-wise quantization to compress weights to 8 bytes, and uses dynamic per-token quantization to compress activations to 8 bytes. 
- Useful for speed ups in high QPS regimes or offline servering on vLLM. 
- Recommended to be used with NVIDIA A100 GPUs. 
#### W8A8-fp8
- Uses channel-wise quantization to compress weights to 8 bytes, and uses dynamic per-token quantization to compress activations to 8 bytes. 
- Useful for speed ups in high QPS regimes or offline servering on vLLM. 
- Recommended to be used with NVIDIA H100 GPUs. 

### Sparsification
Sparsification reduces model complexity by pruning select weight values to zero while retaining essential weights in a subset of parameters. Supported formats include:
* 2:4
    - Uses what algo?

>     - Useful for what case?

Author: leverage sparsify - useful if you want to finetune after

    - Recommended hardware?

## Code

> Oneshot can be carried out using python or cli. 

Reviewer 2: ```suggestion
Oneshot can be invoked using python or cli. 
```
?

### Python 

This is the recommended flow. The example scripts for all the above formats are located in the [examples](../../../examples/) folder. A [W8A8-fp8](../../../examples/quantization_w8a8_fp8/llama3_example.py) example is shown below: 

```python
from transformers import AutoModelForCausalLM, AutoTokenizer

from llmcompressor import oneshot
from llmcompressor.modifiers.quantization import QuantizationModifier

MODEL_ID = "meta-llama/Meta-Llama-3-8B-Instruct"

model = AutoModelForCausalLM.from_pretrained(
    MODEL_ID, device_map="auto", torch_dtype="auto"
)
tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)

recipe = QuantizationModifier(
    targets="Linear", scheme="FP8_DYNAMIC", ignore=["lm_head"]
)

oneshot(model=model, recipe=recipe)
```

### CLI

> Command line argumnets can be carried out under an environment with `llm-compressor` in an editable mode. To install it in editable mode, run:

Reviewer 2: This seems wrong? CLI should be available in non-editable mode as well, editable mode is for local dev
